File: clients/python/proto/response.py
import msgpack


class Response:
    TYPE = 1

    def __init__(
        self,
        msg_id: int,
        error=None,
        result=None
    ):
        self.msg_id = msg_id
        self.error = error
        self.result = result

    def marshal(self) -> bytes:
        return msgpack.packb(
            [
                self.TYPE,
                self.msg_id,
                self.error,
                self.result
            ],
            use_bin_type=False
        )

    @classmethod
    def unmarshal(cls, data: bytes) -> "Response":
        value = msgpack.unpackb(data, raw=False)
        print(repr(value))

        if len(value) != 4:
            raise ValueError("Invalid MessagePack-RPC response")

        msg_type, msg_id, error, result = value

        if msg_type != cls.TYPE:
            raise ValueError("Invalid MessagePack-RPC response type")

        return cls(
            msg_id=msg_id,
            error=error,
            result=result
        )

File: hazectl/proto/response.go
package proto

import (
	"fmt"

	"github.com/vmihailenco/msgpack/v5"
)

type Response struct {
	MsgType uint8
	MsgId   uint32
	Error   any
	Result  any
}

func (r *Response) Init(msgid uint32, err any, result any) {
	r.MsgType = 1
	r.MsgId = msgid
	r.Error = err
	r.Result = result
}

func (r *Response) Marshal() ([]byte, error) {
	return msgpack.Marshal([]any{
		r.MsgType,
		r.MsgId,
		r.Error,
		r.Result,
	})
}

func (r *Response) Unmarshal(data []byte) error {
	var raw []any

	if err := msgpack.Unmarshal(data, &raw); err != nil {
		return err
	}

	if len(raw) != 4 {
		return fmt.Errorf("invalid response: expected 4 elements, got %d", len(raw))
	}

	msgType, ok := raw[0].(uint8)
	if !ok {
		return fmt.Errorf("invalid message type")
	}

	msgId, ok := raw[1].(uint32)
	if !ok {
		return fmt.Errorf("invalid message id")
	}

	r.MsgType = msgType
	r.MsgId = msgId
	r.Error = raw[2]
	r.Result = raw[3]

	return nil
}
